crates/common/fork_choice/src/fork_choice/fork_choice_handlers.rs

```diff
@@ -0,0 +1,217 @@
+        }
+    }
+    root
+}
+
+/// Compute the checkpoint block for epoch ``epoch`` in the chain of block ``root``
+pub fn get_checkpoint_block(store: Store, root: B256, epoch: u64) -> B256 {
+    let epoch_first_slot = compute_start_slot_at_epoch(epoch);
+    get_ancestor(store, root, epoch_first_slot)
+}
+
+/// Update checkpoints in store if necessary
+pub fn update_checkpoints(
+    store: &mut Store,
+    justified_checkpoint: Checkpoint,
+    finalized_checkpoint: Checkpoint,
+) -> anyhow::Result<()> {
+    // Update justified checkpoint
+    if justified_checkpoint.epoch > store.justified_checkpoint.epoch {
+        store.justified_checkpoint = justified_checkpoint
+    }
+
+    // Update finalized checkpoint
+    if finalized_checkpoint.epoch > store.finalized_checkpoint.epoch {
+        store.finalized_checkpoint = finalized_checkpoint
+    }
+    Ok(())
+}
+
+/// Update unrealized checkpoints in store if necessary
+pub fn update_unrealized_checkpoints(
+    mut store: Store,
+    unrealized_justified_checkpoint: Checkpoint,
+    unrealized_finalized_checkpoint: Checkpoint,
+) -> anyhow::Result<()> {
+    // Update unrealized justified checkpoint
+    if unrealized_justified_checkpoint.epoch > store.unrealized_justified_checkpoint.epoch {
+        store.unrealized_justified_checkpoint = unrealized_justified_checkpoint
+    }
+
+    // Update unrealized finalized checkpoint
+    if unrealized_finalized_checkpoint.epoch > store.unrealized_finalized_checkpoint.epoch {
+        store.unrealized_finalized_checkpoint = unrealized_finalized_checkpoint
+    }
+
+    Ok(())
+}
+
+pub fn get_current_store_epoch(store: Store) -> u64 {
+    compute_epoch_at_slot(store.get_current_slot())
+}
+
+pub fn compute_pulled_up_tip(mut store: Store, block_root: B256) -> anyhow::Result<()> {
+    let mut state = store.block_states[&block_root].clone();
+    // Pull up the post-state of the block to the next epoch boundary
+    state.process_justification_and_finalization()?;
+
+    if let Some(justification) = store.unrealized_justifications.get_mut(&block_root) {
+        *justification = state.current_justified_checkpoint;
+    }
+    update_unrealized_checkpoints(
+        store.clone(),
+        state.current_justified_checkpoint,
+        state.finalized_checkpoint,
+    )?;
+
+    // If the block is from a prior epoch, apply the realized values
+    let block_epoch = compute_epoch_at_slot(store.blocks[&block_root].slot);
+    let current_epoch = get_current_store_epoch(store.clone());
+    if block_epoch < current_epoch {
+        store.update_checkpoints(
+            state.current_justified_checkpoint,
+            state.finalized_checkpoint,
+        );
+    }
+
+    Ok(())
+}
+
+/// Run ``on_block`` upon receiving a new block.
+pub async fn on_block(
+    store: &mut Store,
+    signed_block: SignedBeaconBlock,
+    execution_engine: &impl ExecutionApi,
+) -> anyhow::Result<()> {
+    let block = signed_block.message.clone();
```

Comment:
```suggestion
    let block = &signed_block.message;
```

Couldn't this be a reference